pylint/checkers/typecheck.py

```diff
@@ -1902,13 +1902,48 @@ def _detect_unsupported_alternative_union_syntax(self, node: nodes.BinOp) -> Non
                 self._check_unsupported_alternative_union_syntax(node)
 
     def _check_unsupported_alternative_union_syntax(self, node: nodes.BinOp) -> None:
-        """Check if left or right node is of type `type`."""
+        """Check if left or right node is of type `type`.
+
+        If either is, and doesn't
+        support an or operator via a metaclass, infer that this is a mistaken
+        attempt to use alternative union syntax when not supported.
+        """
         msg = "unsupported operand type(s) for |"
-        for n in (node.left, node.right):
-            n = astroid.helpers.object_type(n)
-            if isinstance(n, nodes.ClassDef) and is_classdef_type(n):
-                self.add_message("unsupported-binary-operation", args=msg, node=node)
-                break
+        left_obj = astroid.helpers.object_type(node.left)
+        right_obj = astroid.helpers.object_type(node.right)
+        left_is_type = False
+        right_is_type = False
+
+        def _has_usable_overload(attrs):
+            # We have a candidate overload of the operator. However, if this
+            # is a Python 3.10 runtime but configured for pre-3.10 compatibility
+            # Astroid will have inferred the existence of __or__ / __ror__ on
+            # builtins.type, which is not suitable.
+            is_py310_builtin = all(
```

Comment:
You can just check the `pytype()` of the object. This might help you skirt my nasty test case.
```python
>>> import astroid
>>> l = astroid.extract_node("list")
>>> inferred = next(l.infer())
>>> inferred.pytype()
'builtins.type'
```

Reply:
Thanks, it's good to know that exists. However, it doesn't seem to do what I need here. On a `nodes.ClassDef`, the `pytype()` method will always return `"builtins.type"` so I can't use it to differentiate between a user-defined class an a builtin.